### PR TITLE
Improve ABI conversion

### PR DESCRIPTION
**Pull request by rylev**
*note: since codegen was run for this PR, it is best to review commit by commit*

This improves ABI conversions for COM interfaces:
* Add some docs to the `Abi` trait and clean it up a bit
* Use `::core::mem::MaybeUninit::zeroed` instead of `::core::mem::zeroed` (more on this below)

[`::core::mem::zeroed`](https://doc.rust-lang.org/stable/std/mem/fn.zeroed.html) is soft deprecated because it's hard to use correctly, and  `::core::mem::MaybeUninit::zeroed` is preferred. I think this is reason enough to make the switch, but there's more here to try to straighten out. 

We set all out params to zeros initially. With this change, these out params are considered to be in an uninitialized state when they are passed to the COM method. This is strictly more correct than what we had before where we were using `ManuallyDrop` to control dropping. `MaybeUninit` also does not run the inner type's drop implementation when it is dropped. 

With this change, there is an implicit agreement (which we *must* make explicit) that if the implementor of the COM interface returns an HRESULT which is 0 or above, the out param will be considered initialized. Note that HRESULTs of 0 or above can still be set even if the implementor returns an `Err` (e.g., `Err(Error::OK)`). Because of this, I've added an additional invariant to `Abi` that all implementors of `Abi` must be able to be in a valid state if they are all zeros or they must manually implement `from_abi` and do a check for all zeros and return an `Err` if they are. This check is exactly what `T: Interface` already does.

~~Lastly, as stated above, `MaybeUninit` also does not run the inner type's `drop` implementation when it is dropped. This means we don't need to wrap blittable types in `ManuallyDrop` since this behavior is now controlled by `MaybeUninit`.~~ (we still need to wrap non-blittable types in `ManuallyDrop` because the conversion from type to ABI leads to the type being dropped across FFI boundaries).

### CODE REVIEW

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [inline crates/libs/windows/src/core/abi.rs:48]: Of the impls in this file, I think this is the only one that requires `from_abi` to be unsafe, since it erases the type of the interface. I wonder if we could hang on to the type information? For example, maybe something like this would work?

```rust
unsafe impl<T> Abi for NonNull<T> where Self: Interface {
    type Abi = *const T;
    ...
}
```

A [rough sketch](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cba771cd6dfcb73aa8a029136117652b) of it compiles, although I don't think this works with the rest of the codebase. The main reason is it looks like `Interface` objects are usually structs that wrap an `IUnknown`.

I guess another option is to not have this blanket impl and instead generate one for every interface type we generate. That's probably horrible for compile times though.

Also, I'm guessing most APIs take a void pointer rather than knowing the exact type, so we probably need the type-erased impl anyway.

Anyway, it's probably best to keep the trait and/or the `from_abi` method unsafe for now, but it seems like in theory we could factor things to keep type safety

**Review by eholk — APPROVED** [review]
Looks good! I had a lot of comments around which things are `unsafe` in the `Abi` trait, but I think they are mostly things to think about longer term. I'm fine with merging this as-is.

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:62]: The ABI type is always blittable (e.g. no `Drop`) so that should be fine.

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:26]: It's always safe to convert from a valid ABI type via transmute. Keep in mind that this is only intended for the `windows` crate to use internally and scoped to Windows types that follow a certain ABI convention that makes this safe. 

While a `derive` macro would solve so many problems, the `derive` macros are insanely slow which is why I've scrubbed all but the essentials (`PartialEq` and `Eq` required for constant patterns) out of the `windows` crate. 

Of course, if a certain compiler dev wants to speed up the `derive` macro then I'm all for it. 😄

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs]: No, a COM function/method always `AddRef`s before returning a pointer. So, `from_abi` just assumes ownership of the pointer that has already been `AddRef`d.

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:48]: A big problem I have is not being to have more blanket impls - there's just so much boilerplate code that I'd love to get rid of. Short of derive macros, I'd love to find a way to have more default behavior inheritable. Even with derive macros, this would seem cleaner.

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs:26]: > It's always safe to convert from a valid ABI type via transmute.

This isn't really true. Converting a null pointer to a `T: Interface` is incorrect (and thus why the implementation of `Abi for T where T: Interface` explicitly checks for null), so there's a good reason for there to be an explicit `from_abi` method. 

The reverse (i.e., going from `T` to `T::abi`) is currently always done via transmute. I'd personally be in favor of adding a more explicit `to_abi` method though. This would make the role of the `Abi` trait much clear. It would make verifying safety much more straightforward, and it would eliminate some calls to `std::mem::transmute` which I think is a goal we should always have. 

So, I agree that a `derive` macro probably isn't worth implementation complexity and resulting compile time slow down, but I do think it would be worth writing a *little* bit more code that makes all of this much more straightforward to understand.

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs:16]: I like this idea a lot. This also has the potential to help us address the issue of what is and isn't allowed across FFI boundaries. Right now, for example, when users declare COM interfaces, they are just expected to use FFI safe types - we don't check for them. We could, however, consider putting in type checks that any type used in a COM interface method signature implements `AbiSafe` which could make implementing COM interfaces harder to mess up. 

This is probably outside the scope of this PR however. 

> What are the requirements for being FFI-safe? Is #[repr(C)] sufficient or necessary?

FFI-safe in this context is somewhat nebulous. A type that is `#[repr(C)]` or some other stable representation (or `#[repr(transparent)]` over such a representation) is necessary but not sufficient. It also needs to not have a `Drop` impl as we often pass such values by value over an FFI boundary but we don't want their destructors to run. For example, `HSTRING` has a stable representation, but it also has a `Drop` impl that decrements a ref count (and potentially frees memory). So `HSTRING` isn't really FFI-safe for our purposes, but `ManuallyDrop<HSTRING>` is.

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs:25]: Yes, unfortunately this is a case where `from_abi` is often times safe (especially when `Self::Abi == Self`, but not always (as you've pointed out in the case of `Abi` being a pointer - and especially a void pointer).

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs]: I came to that conclusion that @kennykerr laid out here, and I thought I had replaced this comment with an explanation as such, but I guess I was dreaming... I'll fix this.

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs:48]: Even if we have the interface type in `Abi` this would still have to be unsafe. Reading raw pointers is always unsafe since so many invariants are not upheld by the compiler, and it is up to the code that transforms a raw pointer to a more concrete Rust type to ensure those invariants have been upheld. For example:

*raw pointers are not type safe (i.e., casting raw pointers is "safe")*

```rust
let ptr: *const usize = &42usize as *const usize;
let ptr = ptr as *const IUknown::Vtable; // this line is wrong but not considered "unsafe"
// It is up to the next line to make sure `ptr` is well formed (which it is not)
unsafe { IUnknown::from_abi(ptr) }
```

File: crates/libs/windows/src/core/abi.rs
Comment (rylev) [reply crates/libs/windows/src/core/abi.rs]: I wasn't dreaming - I didn't notice this comment was made on a specific commit. This line was removed in a later commit. Sorry!

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:26]: > This isn't really true.

Sorry if that was misleading. I was giving the short-hand C++ explanation. The Rust explanation is a little more nuanced only because we distinguish between nullable types and non-nullable types which we don't do in C++. You'll see this in `RuntimeType::DefaultType`. So in Rust, *it's always safe to convert from a valid ABI type via transmute to a corresponding "default type"*. I do need to merge/harmonize the `Abi` trait and the `RuntimeType` trait so let's avoid overhauling the `Abi` trait without keeping that in mind. Happy to consider pairing `to_abi` and `from_abi` - the confusion just comes from the split between the `Abi` and `RuntimeType` but I think it will be a lot simpler to understand once those are combined.

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:25]: It should always be safe once we harmonize `Abi` and `RuntimeType`.

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [reply crates/libs/windows/src/core/abi.rs:26]: Just for my own understanding, what is `RuntimeType` used for now? What do you envision things looking like after merging/harmonizing `Abi` and `RuntimeType`?

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [reply crates/libs/windows/src/core/abi.rs:64]: I think in this case both `transmute` and `transmute_copy` end up doing the same thing, so either one is fine. I personally think `transmute` is better stylistically because we do not need or even want to use `abi` after transmuting it.

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [reply crates/libs/windows/src/core/abi.rs]: Oh, sorry! That's what I get for reviewing this commit by commit.

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [reply crates/libs/windows/src/core/abi.rs:48]: Ah, good point. The fact that casting between different unsafe pointer traits is safe means `from_abi` has to be unsafe.

File: crates/libs/windows/src/core/abi.rs
Comment (eholk) [reply crates/libs/windows/src/core/abi.rs:16]: > This is probably outside the scope of this PR however.

Yeah, for sure!

Thanks for the explanation about what we mean by FFI-safe. If we move towards having an `AbiSafe` trait in the future, that'd be good to include in the documentation for that trait.

**Review by eholk — APPROVED** [review]

File: crates/libs/windows/src/core/abi.rs
Comment (kennykerr) [reply crates/libs/windows/src/core/abi.rs:26]: `RuntimeType` was originally required beyond just `Abi` because WinRT includes generic COM interfaces where certain method signatures may have different types and semantics based on the generic type parameters and if those type parameters are nullable then need to know what the default type ends up being in order to build the projection correctly. It's all rather convoluted and more so in Rust because of `Option`/`NonNull`. Then throw ABI safety with nullable types and the need for a memory layout compatible borrow type it makes sense to harmonize this, which is what I've been meaning to do for some time. So let's hold off on any major `Abi` trait changes until I've got that sorted out and then I can get you guys to look it over.

**Review by kennykerr — APPROVED** [review]

**Pull request merged**
